test/jdk/jdk/internal/misc/TerminatingThreadLocal/TestTerminatingThreadLocal.java

```diff
@@ -30,47 +30,67 @@
-
-        ttlTestVirtual(666, ThreadLocal::get, 666);
+    @SuppressWarnings("unchecked")
+    static <T> Object[] testCase(T v0, Consumer<? super TerminatingThreadLocal<T>> action, T... v1) {
+        return new Object[] { v0, action, v1 };
     }
 
-    static <T> void ttlTestSet(T v0, T v1) {
-        ttlTestPlatform(v0, ttl -> {                                         }    );
-        ttlTestPlatform(v0, ttl -> { ttl.get();                              }, v0);
-        ttlTestPlatform(v0, ttl -> { ttl.get();   ttl.remove();              }    );
-        ttlTestPlatform(v0, ttl -> { ttl.get();   ttl.set(v1);               }, v1);
-        ttlTestPlatform(v0, ttl -> { ttl.set(v1);                            }, v1);
-        ttlTestPlatform(v0, ttl -> { ttl.set(v1); ttl.remove();              }    );
-        ttlTestPlatform(v0, ttl -> { ttl.set(v1); ttl.remove(); ttl.get();   }, v0);
-        ttlTestPlatform(v0, ttl -> { ttl.get();   ttl.remove(); ttl.set(v1); }, v1);
+    @DataProvider
+    public Object[][] testCases() {
+        Integer NULL = null;
+        return new Object[][] {
+            testCase(42, ttl -> {                                             }     ),
+            testCase(42, ttl -> { ttl.get();                                  }, 42 ),
+            testCase(42, ttl -> { ttl.get();    ttl.remove();                 }     ),
+            testCase(42, ttl -> { ttl.get();    ttl.set(112);                 }, 112),
+            testCase(42, ttl -> { ttl.set(112);                               }, 112),
+            testCase(42, ttl -> { ttl.set(112); ttl.remove();                 }     ),
+            testCase(42, ttl -> { ttl.set(112); ttl.remove(); ttl.get();      }, 42 ),
+            testCase(42, ttl -> { ttl.get();    ttl.remove(); ttl.set(112);   }, 112),
+
+            testCase(NULL, ttl -> {                                           }      ),
+            testCase(NULL, ttl -> { ttl.get();                                }, NULL),
+            testCase(NULL, ttl -> { ttl.get();    ttl.remove();               }      ),
+            testCase(NULL, ttl -> { ttl.get();    ttl.set(112);               }, 112 ),
+            testCase(NULL, ttl -> { ttl.set(112);                             }, 112 ),
+            testCase(NULL, ttl -> { ttl.set(112); ttl.remove();               }      ),
+            testCase(NULL, ttl -> { ttl.set(112); ttl.remove(); ttl.get();    }, NULL),
+            testCase(NULL, ttl -> { ttl.get();    ttl.remove(); ttl.set(112); }, 112 ),
+
+            testCase(42, ttl -> { ttl.get();     ttl.set(NULL);               }, NULL),
+            testCase(42, ttl -> { ttl.set(NULL);                              }, NULL),
+            testCase(42, ttl -> { ttl.set(NULL); ttl.remove();                }      ),
+            testCase(42, ttl -> { ttl.set(NULL); ttl.remove(); ttl.get();     }, 42  ),
+            testCase(42, ttl -> { ttl.get();     ttl.remove(); ttl.set(NULL); }, NULL),
+        };
```

Comment:
Hm, what about expanding the test cases programmatically. Like this?
https://github.com/plevart/jdk/commit/ffc681f3cccfdf473d4cc037e5f3e50f8172eec3
...it's easier to follow. Some nits included...

Reply:
> Hm, what about expanding the test cases programmatically. Like this?
> [plevart@ffc681f](https://github.com/plevart/jdk/commit/ffc681f3cccfdf473d4cc037e5f3e50f8172eec3)
> ...it's easier to follow. Some nits included...

I'm not sure that it's any easier to follow but I don't mind merging that in. The other cleanup is good.